src/Services/Notification.vala

```diff
@@ -56,67 +53,56 @@ public class Notifications.Notification : Object {
     private const string X_CANONICAL_PRIVATE_KEY = "x-canonical-private-synchronous";
     private const string DESKTOP_ENTRY_KEY = "desktop-entry";
     private const string FALLBACK_DESKTOP_ID = "gala-other" + DESKTOP_ID_EXT;
-    private bool pid_acquired;
 
-    public Notification.from_message (DBusMessage message, uint32 _id) {
+    public Notification.from_message (DBusMessage message, uint32 id) {
         var body = message.get_body ();
-
-        data_session = false;
-
-        app_name = get_string (body, Column.APP_NAME);
-        app_icon = get_string (body, Column.APP_ICON);
-        summary = get_string (body, Column.SUMMARY);
-        message_body = get_string (body, Column.BODY);
         hints = body.get_child_value (Column.HINTS);
-        expire_timeout = get_int32 (body, Column.EXPIRE_TIMEOUT);
-        replaces_id = get_uint32 (body, Column.REPLACES_ID);
-        id = _id;
-        sender = message.get_sender ();
-
-        actions = body.get_child_value (Column.ACTIONS).dup_strv ();
-        timestamp = new GLib.DateTime.now_local ();
-        unix_time = timestamp.to_unix ();
 
-        desktop_id = lookup_string (hints, DESKTOP_ENTRY_KEY);
-        if (desktop_id != "" && !desktop_id.has_suffix (DESKTOP_ID_EXT)) {
-            desktop_id += DESKTOP_ID_EXT;
+        Object (
+            actions: body.get_child_value (Column.ACTIONS).dup_strv (),
+            app_icon: get_string (body, Column.APP_ICON),
+            app_name: get_string (body, Column.APP_NAME),
+            desktop_id: lookup_string (hints, DESKTOP_ENTRY_KEY),
+            id: id,
+            message_body: get_string (body, Column.BODY),
+            replaces_id: get_uint32 (body, Column.REPLACES_ID),
+            sender: message.get_sender (),
+            summary: get_string (body, Column.SUMMARY),
+            timestamp: new GLib.DateTime.now_local ()
+        );
+    }
 
-            app_info = new DesktopAppInfo (desktop_id);
-        }
+    public Notification.from_data (
+        uint32 id, string app_name, string app_icon,
+        string summary, string message_body, string[] _actions,
+        string desktop_id, int64 unix_time, string sender
+    ) {
+        Object (
+            actions: actions,
+            app_icon: app_icon,
+            app_name: app_name,
+            data_session: true,
+            desktop_id: desktop_id,
+            id: id,
+            message_body: message_body,
+            replaces_id: 0,
+            sender: sender,
+            summary: summary,
+            timestamp: new GLib.DateTime.from_unix_local (unix_time)
+        );
+    }
 
-        if (app_info == null) {
+    construct {
+        if (desktop_id != "") {
+            // Avoid example.desktop.desktop
+            desktop_id.replace (".desktop", "");
```

Comment:
replace it with either a regex that makes sure that we are only replacing the end of the desktop_id or do something like:
```vala
if desktop_id.has_suffix (".desktop") {
    desktop_id = desktop_id.substring (0, - ".desktop".length)
}
```

Reply:
But we are adding ".desktop" to the end again in the next line??  It looks like the intention was to remove multiple occurrences of ".desktop" and replace with just one at the end.

Could use DESKTOP_ID_EXT rather than ".desktop"